File: src/lib/scoring.ts
// Frontend display helpers for the Agent Report and star labels. The numeric
// star scores themselves (current_score / current_stars / potential_*) are
// computed once on the backend (roster_service._compute_star_scores) and
// consumed from the API — this module holds only presentation-layer
// derivations, so the score algorithm has a single source of truth.

function pct(r: any): number { return Number(r?.pct ?? 0) }

export function pillarScores(s: any, w: any) {
  const ringVals = [pct(s.brawl), pct(s.puroresu), pct(s.hardcore), pct(s.technical), pct(s.air), pct(s.flash)]
  return {
    primary: Math.max(...ringVals),
    perf: calcPerformance(s),
    pop: w.pop?.pct ?? 0,
    fund: (pct(s.psych) + pct(s.basics) + pct(s.selling) + pct(s.consistency) + pct(s.safety)) / 5,
    stamina: pct(s.stamina),
  }
}

export function calcPerformance(skills: any): number {
  if (!skills) return 0
  const cha = Number(skills.charisma?.pct ?? 0)
  const mic = Number(skills.mic?.pct ?? 0)
  const act = Number(skills.acting?.pct ?? 0)
  const bestVis = Math.max(Number(skills.star?.pct ?? 0), Number(skills.looks?.pct ?? 0), Number(skills.menace?.pct ?? 0))
  return (cha + mic + act + bestVis) / 4
}

export function relTier(cp: number) {
  const c = Math.max(cp, 10)
  return { elite: c + 25, strong: c + 15, solid: c + 5, weak: c - 10, poor: c - 20 }
}

export function proImpact(stat: number, threshold: number, cp: number): number {
  const gap = stat - threshold
  const scale = Math.max(cp * 0.12, 5)
  return Math.max(0.05, Math.min(5, 1 + gap / scale))
}

export function conSeverity(stat: number, threshold: number, cp: number): number {
  const gap = threshold - stat
  const scale = Math.max(cp * 0.12, 5)
  return Math.max(0.05, Math.min(5, 1 + gap / scale))
}

export function isElite(stat: number) { return stat >= 90 }
export function isWorldClass2(...stats: number[]) { return stats.filter(s => s >= 90).length >= 2 }
export function isWrestler(w: any) {
  if (w.retired) return false
  const p = w.positions || []; return p.includes('Wrestler') || p.includes('Occasional')
}

export function detectStyle(s: any, w: any): string {
  const pct = (r: any) => Number(r?.pct ?? 0)
  const age = w.age
  const b = pct(s.brawl); const pu = pct(s.puroresu); const ha = pct(s.hardcore)
  const te = pct(s.technical); const ai = pct(s.air); const fl = pct(s.flash)
  const ch = pct(s.charisma); const mi = pct(s.mic); const st = pct(s.star)
  const ps = pct(s.psych); const ba = pct(s.basics); const se = pct(s.selling)
  const co = pct(s.consistency); const sa = pct(s.safety); const stm = pct(s.stamina)
  const at = pct(s.athletic); const po = pct(s.power); const to = pct(s.toughness)
  const ex = pct(s.experience); const mn = pct(s.menace); const lk = pct(s.looks)
  const allRing = [b, pu, ha, te, ai]
  const ringVals = [...allRing].sort((x, y) => y - x)
  const primaryScore = (ringVals[0]*0.50 + ringVals[1]*0.25 + ringVals[2]*0.15 + ringVals[3]*0.07 + ringVals[4]*0.03) * (1 + Math.min(5, Math.max(0, (fl - 50) * 0.1)) / 100)
  const entScore = (ch + mi + st) / 3
  const TIER = (t: string) => {
    if (['Monster','Dominator','Ring General','Complete','Superstar'].includes(t)) return 30
    if (['Well-Rounded','Solid Hand','All-Rounder'].includes(t)) return 25
    if (['Technician','Bruiser','High-Flyer','Powerhouse','Brawler','Enforcer','Specialist','Entertainer','Showman'].includes(t)) return 15
    return 0
  }
  const candidates: [string, number][] = []
  if (po > 70 && mn > 75 && to > 65 && ai < 60 && te < 60 && (b + pu + ha) / 3 < 65)
    candidates.push(['Monster', (po + mn + to) / 3 + TIER('Monster')])
  if (po > 80 && at > 75 && to > 75 && b > 70 && te < 70)
    candidates.push(['Dominator', (po + at + to + b) / 4 + TIER('Dominator')])
  if (ps > 74 && ex >= 100 && se > 70 && sa > 80 && co > 80 && entScore < 75)
    candidates.push(['Ring General', (ps + se + sa + co) / 4 + TIER('Ring General')])
  if (ringVals.slice(0, 3).every(v => v >= 55) && ps >= 85 && ba >= 80 && co >= 80 && sa >= 70 && stm >= 75 && entScore >= 65)
    candidates.push(['Complete', ((ringVals[0] + ringVals[1] + ringVals[2] + ringVals[3]) / 4 * 0.5 + (ps + ba + co + sa) / 4 * 0.3 + entScore * 0.2) + TIER('Complete')])
  if (st >= 90 && lk >= 80 && po >= 75 && entScore >= 75 && ps >= 70)
    candidates.push(['Superstar', (entScore * 0.3 + st * 0.3 + lk * 0.2 + po * 0.2) + TIER('Superstar')])
  if (entScore > primaryScore && entScore >= 76)
    candidates.push(['Entertainer', entScore + TIER('Entertainer')])
  if (b > 65 && pu > 60 && ha > 50)
    candidates.push(['Bruiser', (b + pu + ha) / 3 + TIER('Bruiser')])
  if (te > 75 && ba > 70 && ai < 65)
    candidates.push(['Technician', (te + ba) / 2 + TIER('Technician')])
  if ((ai >= 70 || fl >= 70) && entScore >= 70 && st >= 75 && se >= 75)
    candidates.push(['Showman', (ai + fl + se + ch) / 4 + TIER('Showman')])
  if (ai > 75 && at > 70 && fl > 65)
    candidates.push(['High-Flyer', (ai + at + fl) / 3 + TIER('High-Flyer')])
  if (po > 75 && b > 60)
    candidates.push(['Powerhouse', (po + b) / 2 + TIER('Powerhouse')])
  if (b > 70 && to > 60)
    candidates.push(['Brawler', (b + to) / 2 + TIER('Brawler')])
  if (to > 70 && po > 65 && mn > 60)
    candidates.push(['Enforcer', (to + po + mn) / 3 + TIER('Enforcer')])
  if (ringVals.slice(0, 3).every(v => v >= 60) && (ba + co + sa) / 3 > 70 && entScore > 60)
    candidates.push(['Well-Rounded', (allRing.reduce((a, b) => a + b, 0) / 5 + entScore) / 2 + TIER('Well-Rounded')])
  if (allRing.filter(v => v >= 65 && v <= 79).length >= 3 && co > 70 && sa > 70)
    candidates.push(['Solid Hand', allRing.reduce((a, b) => a + b, 0) / 5 + TIER('Solid Hand')])
  if (Math.max(...allRing) >= 85 && allRing.filter(v => v >= 60).length === 1)
    candidates.push(['Specialist', Math.max(...allRing) + TIER('Specialist')])
  if (age >= 38 && ex >= 90 && ps > 70 && se > 70 && (at < 60 || stm < 60))
    candidates.push(['Veteran', (ex + ps + se) / 3])
  if (age <= 25 && allRing.every(v => v >= 40 && v <= 65))
    candidates.push(['Young Lion', allRing.reduce((a, b) => a + b, 0) / 5])
  if (allRing.filter(v => v >= 70).length >= 3 && Math.max(...allRing) < 90)
    candidates.push(['All-Rounder', allRing.reduce((a, b) => a + b, 0) / 5])
  if (candidates.length === 0) return ''
  candidates.sort((a, b) => b[1] - a[1])
  return candidates[0][0]
}

export function workerType(w: any): string {
  if (!w.skills) return ''
  return detectStyle(w.skills, w)
}

export function agePrefix(age: number): string {
  if (age >= 44) return 'Aging'
  if (age >= 41) return 'Grizzled'
  if (age >= 38) return 'Veteran'
  if (age >= 35) return 'Seasoned'
  if (age >= 32) return 'Established'
  if (age >= 23) return 'Up-and-Coming'
  if (age >= 20) return 'Rising'
  if (age < 20) return 'Young'
  return ''
}

function isBangedUp(w: any): boolean {
  const p = w.physical
  if (!p) return false
  return Math.min(p.condition1 ?? 100, p.condition2 ?? 100, p.condition3 ?? 100, p.condition4 ?? 100) < 55
}

export function starLabel(w: any, stars: number, _isPotential: boolean, score?: number): string {
  if (!isWrestler(w)) {
    const pos: string[] = w.positions || []
    const levels = ['Ineffective', 'Poor', 'Below Average', 'Average', 'Good', 'Very Good', 'Impressive', 'Exceptional', 'World Class']
    const roleLabels: Record<string, string> = {
      Referee: 'Referee', Announcer: 'Announcer', Colour: 'Colour Commentator',
      Manager: 'Manager', Personality: 'Personality', 'Road Agent': 'Agent',
    }
    const role = pos.find(p => roleLabels[p]) || 'Announcer'
    const label = roleLabels[role] || 'Announcer'
    const sv = score ?? (stars * 10)
    const idx = sv >= 90 ? 8 : sv >= 80 ? 7 : sv >= 70 ? 6 : sv >= 57.5 ? 5 : sv >= 40 ? 4 : sv >= 20 ? 3 : sv >= 10 ? 2 : sv >= 5 ? 1 : 0
    return `${levels[idx]} ${label}`
  }
  const st = workerType(w)
  const s = score ?? (stars * 10)
  if (!w.retired && isBangedUp(w) && st && !['Complete', 'Well-Rounded'].includes(st)) return `Banged Up ${st}`

  const ap = agePrefix(w.age)
  const isYoung = ap === 'Young' || ap === 'Rising' || ap === 'Up-and-Coming'
  const isOld = ap === 'Veteran' || ap === 'Grizzled' || ap === 'Aging'

  const adjForm: Record<string, string> = {
    Monster: 'Monstrous', Technician: 'Technical', Bruiser: 'Bruising',
    'High-Flyer': 'High-Flying', Powerhouse: 'Powerful', Brawler: 'Brawling',
    Enforcer: 'Enforcing', Entertainer: 'Entertaining', Specialist: 'Specialist',
    'Solid Hand': 'Solid', Veteran: 'Veteran', 'Young Lion': 'Young Lion',
    'All-Rounder': 'All-Rounding', 'Ring General': 'Ring General',
    Complete: 'Complete', 'Well-Rounded': 'Well-Rounded',
  }

  const fmtPos = (posAdj: string, posNoun: string, type: string | null) => {
    if (!type) return posNoun
    const typeFirst = ['Complete', 'Well-Rounded']
    if (typeFirst.includes(type)) return `${type} ${posNoun}`
    return `${posAdj} ${type}`
  }

  if (s >= 70) return fmtPos('Main Event', 'Main Eventer', st)
  if (s >= 57.5) {
    if (isYoung) return st ? `Rising ${st}` : 'Rising Upper Midcarder'
    if (isOld) return st ? (adjForm[st] !== st ? `${adjForm[st]} Veteran` : `Veteran ${st}`) : 'Veteran Upper Midcarder'
    return fmtPos('Upper Midcard', 'Upper Midcarder', st)
  }
  if (s >= 40) {
    if (isYoung) return st ? `Rising ${st}` : 'Rising Midcarder'
    if (isOld) return st ? `Established ${st}` : 'Established Midcarder'
    return fmtPos('Midcard', 'Midcarder', st)
  }
  if (s >= 20) {
    if (isOld) return st ? `Journeyman ${st}` : 'Journeyman'
    return st ? `Preliminary ${st}` : 'Preliminary'
  }
  if (!_isPotential && (w.potential_stars ?? 0) >= 2.5) return 'Developing Young Lion'
  const lowLabel = w.age >= 30 ? 'Deadwood' : w.age >= 28 ? 'Enhancement Talent' : 'Developing'
  return st ? `${lowLabel} ${st}` : lowLabel
}

File: python/domains/worker/scoring.py
"""Star-rating scoring engine — port of the frontend logic (src/lib/scoring.ts)
so ratings are computed once on the backend and identical across list/detail
views. Pure with respect to its Worker argument: never touches the store,
only fields already set on `w` (see assembly.py for how those get set)."""

from models import Worker
from .labels import detect_style, age_prefix, is_banged_up, usage_label


def _compute_star_scores(w: Worker):
    s = w.skills
    if not s:
        w.current_score = 0; w.potential_score = 0
        w.current_stars = 0.5; w.potential_stars = 0.5
        return

    def _pct(k): return (getattr(s, k).pct if hasattr(getattr(s, k, None), 'pct') else 0) or 0

    is_wrestler = 'Wrestler' in (w.positions or []) or 'Occasional' in (w.positions or [])
    if not is_wrestler:
        pop = w.pop.pct if w.pop else 0
        pos = w.positions or []
        if 'Referee' in pos:
            skills = [_pct('refereeing'), _pct('respect'), pop]
        elif 'Announcer' in pos:
            skills = [_pct('announcing'), _calc_perf(s), pop]
        elif 'Colour' in pos:
            skills = [_pct('colour'), _calc_perf(s), pop]
        elif 'Manager' in pos or 'Personality' in pos:
            skills = [_calc_perf(s), pop]
        elif 'Road Agent' in pos:
            skills = [_pct('psych') * 0.85 + _pct('experience') * 0.03 + _pct('respect') * 0.12]
        else:
            skills = []
        if not skills:
            w.current_score = 0; w.potential_score = 0
            w.current_stars = 0.5; w.potential_stars = 0.5
            return
        avg = sum(skills) / len(skills)
        w.current_score = round(max(0, min(100, avg)))
        w.potential_score = w.current_score
        w.current_stars = _stars_from_score(w.current_score)
        w.potential_stars = w.current_stars
        return

    rv = [_pct(k) for k in ('brawl', 'puroresu', 'hardcore', 'technical', 'air')]
    fl = _pct('flash')
    primary = max(rv)
    perf = _calc_perf(s)
    pop = w.pop.pct if w.pop else 0
    fund = (_pct('psych') + _pct('basics') + _pct('selling') + _pct('consistency') + _pct('safety')) / 5
    stamina = _pct('stamina')
    psych = _pct('psych')
    best_skill = max(primary, perf)
    worst_skill = min(primary, perf)
    secondary = (psych + fund + stamina) / 3
    worker_level = pop * 0.50 + best_skill * 0.25 + worst_skill * 0.15 + secondary * 0.10

    core85 = sum(1 for k in ('charisma', 'mic', 'acting') if _pct(k) >= 85)
    core90 = sum(1 for k in ('charisma', 'mic', 'acting') if _pct(k) >= 90)
    best_vis85 = max(_pct('star'), _pct('looks'), _pct('menace')) >= 85
    if core85 >= 3: worker_level += 10
    elif core90 >= 2: worker_level += 10
    elif core85 >= 2: worker_level += 5
    if best_vis85 and core85 >= 2: worker_level += 5
    worker_level += max(-10, min(10, _attr_modifier(w)))

    company_pop = w.company_area_pop or 0
    roster_avg_pop = w.roster_avg_pop or 0
    company_level = max(company_pop, roster_avg_pop) * 0.65 + min(company_pop, roster_avg_pop) * 0.35 if roster_avg_pop > 0 else company_pop
    delta = worker_level - company_level
    score = 60 + delta * 1.5
    pop_gap_current = max(0, company_level - pop)
    if pop_gap_current > 10:
        score -= pop_gap_current * 0.6

    if stamina < 60:
        score -= (60 - stamina) * 0.4

    flash_buff = min(5, max(0, (fl - 50) * 0.1))
    score += flash_buff

    phys = w.physical
    if phys:
        lowest = min(phys.condition1, phys.condition2, phys.condition3, phys.condition4)
        if lowest < 55:
            score -= (55 - lowest) * 0.35

    if delta < 0:
        rp = w.roster_avg_primary or 0
        re = w.roster_avg_ent or 0
        if rp > 0 or re > 0:
            roster_level = rp * 0.35 + re * 0.35 + (w.roster_avg_psych or 0) * 0.10 + (w.roster_avg_fund or 0) * 0.07
            roster_delta = worker_level - roster_level
            if roster_delta > 0:
                score += min((roster_delta / max(company_pop, 1)) * 15, 15)

    score = max(0, min(100, score))
    w.current_score = round(score)
    w.current_stars = _stars_from_score(score)

    skill_level = best_skill * 0.50 + worst_skill * 0.35 + secondary * 0.15 + max(-10, min(10, _attr_modifier(w)))
    skill_level = max(0, min(100, skill_level))
    skill_delta = skill_level - company_level
    potential = max(score + _age_growth(w.age), min(100, 60 + skill_delta * 1.5))
    potential = max(80 if core85 >= 2 and w.age <= 30 else 0, min(100, potential))
    if w.age > 42:
        potential = min(potential, score)
    w.potential_score = round(potential)
    w.potential_stars = _stars_from_score(potential)

    # Derived fields for frontend consumption (no more frontend computation)
    w.is_wrestler = is_wrestler
    w.pillar_primary = round(primary)
    w.pillar_perf = round(perf)
    w.pillar_pop = pop
    w.perf_score = round(_calc_perf(s))
    w.worker_type = detect_style(w)
    w.age_prefix = age_prefix(w.age)
    w.is_banged_up = is_banged_up(w)
    w.usage_label = usage_label(w, w.current_stars, w.current_score)
    w.potential_usage_label = usage_label(w, w.potential_stars, w.potential_score, is_potential=True)


def _calc_perf(skills) -> float:
    def p(k): return (getattr(skills, k).pct if hasattr(getattr(skills, k, None), 'pct') else 0) or 0
    cha, mic, act = p('charisma'), p('mic'), p('acting')
    best_vis = max(p('star'), p('looks'), p('menace'))
    return (cha + mic + act + best_vis) / 4


def _attr_modifier(w: Worker) -> float:
    attrs = (w.attributes or []) if hasattr(w, 'attributes') else (getattr(w, 'attributes', None) or [])
    if not attrs:
        attrs = []
    has = lambda i: i in attrs
    mod = 0.0
    age = w.age or 0
    if age <= 20: mod -= 2
    elif age <= 22: mod -= 1
    elif age <= 25: mod += 1
    elif age <= 28: mod += 2
    elif age <= 31: mod += 2
    elif age <= 34: mod += 1
    elif age <= 37: mod += 0
    elif age <= 40: mod -= 1
    elif age <= 43: mod -= 3
    else: mod -= 5
    if has(507): mod += 2
    if has(509): mod += 1
    if has(510): mod -= 2
    pos_pers = {1: 3, 3: 2, 4: 2, 5: 3, 8: 4, 9: 5, 11: 1}
    neg_pers = {12: -2, 13: -1, 14: -1, 15: -4, 16: -1, 17: -5, 18: -2, 19: -1, 20: -1, 21: -3, 22: -2, 23: -2, 24: -1, 25: -1, 26: -5, 27: -3, 28: -5}
    pers = next((id for id in attrs if 1 <= id <= 28), None)
    if pers and pers in pos_pers: mod += pos_pers[pers]
    elif pers and pers in neg_pers: mod += neg_pers[pers]
    if has(548): mod += 3
    elif has(547): mod += 2
    elif any(has(i) for i in [225, 226, 227, 228, 229, 231, 232, 233, 535, 550]): mod += 1
    if has(314) or has(315): mod += 2
    if has(310) or has(313): mod -= 2
    if any(has(i) for i in [125, 131, 134]): mod += 1
    if has(346): mod += 1
    if has(345): mod += 1
    if has(348): mod += 1
    if has(352): mod += 1
    if has(502): mod += 1
    if has(103): mod += 2
    if has(104) or has(105): mod += 1
    if has(122): mod += 1
    if has(106): mod += 1
    if has(347): mod -= 2
    if has(344): mod -= 1
    if has(118): mod -= 2
    if has(119): mod -= 1
    if has(351): mod -= 2
    if any(has(i) for i in [340, 341, 374, 375]): mod -= 1
    if has(545): mod -= 2
    if has(546): mod -= 1
    if has(543): mod -= 5
    if has(544): mod -= 1
    if has(349): mod -= 1
    if has(353): mod -= 1
    danger = [197, 198, 199, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 27, 563]
    danger_count = sum(1 for i in danger if has(i))
    if danger_count: mod -= danger_count * 2
    if any(has(i) for i in [520, 521, 522, 523, 524]): mod -= 2
    if has(552): mod -= 2
    perception = getattr(getattr(w, 'contract', None), 'perception', 0) or 0
    if perception == 1: mod += 4
    elif perception == 2: mod += 2
    return max(-35, min(45, round(mod * 0.5)))


def _stars_from_score(score: float) -> float:
    if score >= 90: return 5
    if score >= 80: return 4.5
    if score >= 70: return 4
    if score >= 60: return 3.5
    if score >= 50: return 3
    if score >= 40: return 2.5
    if score >= 30: return 2
    if score >= 20: return 1.5
    if score >= 10: return 1
    return 0.5


def _age_growth(age: int) -> float:
    if age <= 20: return 15
    if age <= 22: return 12
    if age <= 25: return 10
    if age <= 28: return 7
    if age <= 31: return 5
    if age <= 34: return 3
    if age <= 37: return 0
    if age <= 40: return -3
    if age <= 43: return -5
    return -8
